### ISSUE TITLE
Improve network and runtime performance

### ISSUE THREAD

**Issue by forabi** [hollowverse-archive/hollowverse#111]
## Rationale
The app is currently bundled as a single 800KB JS bundle, that contains CSS styles, routes and all vendor libraries.

So hollowverse is designed as mobile-only web app. We need to optimize for this use case. Mobile devices have different performance characteristics compared to desktop computers. Desktop computers today are powerful and have plenty of computing power. On the other hand, mobile devices (especially Android devices, which comprise at least 80% of mobile devices today) have limited CPU resources. The single-core performance of the top high-end Android devices is way below a typical desktop computer. Single-core performance is probably the most relevant performance metric for JavaScript apps, considering the single-threaded nature of the language.

JavaScript needs to be:
* Fetched
* Parsed
* Compiled

So loading 800KB of JavaScript not only impacts the network time of the app (time to first byte), but it also delays time to first meaningful paint, which might be the important performance metric for mobile web apps. Users expect some content to be displayed within 3 seconds of first visit.

To put that into perspective, consider [the following chart](https://mobile.twitter.com/HenriHelvetica/status/877924754195324928):

> Parse times of 1MB of #javascript of various devices as mentioned by @tkadlec #FluentConf 
> ![dc8d2xbxuaitnnq](https://user-images.githubusercontent.com/1012761/27686625-4bd3a1c2-5cdc-11e7-8da4-37f77998b806.jpg)

So we are pretty close to 1MB of JS. To make matters worse, using CSS-in-JS means that styles are embedded in JavaScript. So instead of fetching the styles and JS in parallel, the browser has to wait until JS is fetched, parsed and compiled, additionally delaying time to first paint.

Here is how our current production bundle breaks down:

![screenshot_20170629_144916](https://user-images.githubusercontent.com/1012761/27687116-3b5bfe96-5cde-11e7-8c87-7431693f8650.png)

[`stats.json` file](https://www.dropbox.com/s/oh8tefig4xrczrl/stats.json?dl=0) can be viewed using https://chrisbateman.github.io/webpack-visualizer/.

* 96% of our code is vendor libraries.
* 24% of our code is React and React DOM.
* We are bundling the entire lodash library, even though we are only including the following functions: `pick`, `pickBy` and `sortBy`.
* We are including a huge number of polyfills by requiring `babel-polyfill`, even if not all of them are required. Browsers have [pretty good support for ES6](https://kangax.github.io/compat-table/es6/).
* We are using CSS-in-JS.
* We are bundling all the routes that the app is made of, even those that are not needed for the home page or may never even be visited by the users. So are loading and parsing a lot of JavaScript that will probably not be used at all.
* Although not included in the bundle, we are loading the entire font-awesome CSS. We may not need more than a couple icons from that font.

## What we can do to improve
This is based on my recent experience rewriting [WebCeph](https://github.com/forabi/WebCeph) and another private project for a client.

So WebCeph 1 was a 600KB monolithic bundle of minified JS that included React, React DOM, Material UI (which uses inline styles) and all other vendor libraries. I managed to cut the initial bundle size significantly by:

* Switching to Preact: While `react` may not be very large (36K out of 800K), `react-dom` is very huge (167K) it uses a [synthetic event system](https://facebook.github.io/react/docs/events.html) which re-implements a lot of built-in browser functionality. Preact uses the native browser event system, and is only 3kb. `preact-compat` provides a compatibility layer for React libraries. I had a project built with React, Redux and React Router 4 and was almost finished. I managed to switch to Preact without having to change a single line of code. `preact` and `preact-compat` combined are only 12KB minified. This is a very huge win for free!
* Extracting vendor code into separate Webpack chunks: using the `CommonsChunksPlugin` for webpack, I extracted all Preact libraries into a separate chunk that can be fetched in parallel with my app code. This is also better for long term cashing when used with hashed file names, as changing our own app code won't cause the browser to re-download the vendor code.
* Going back to plain ol' CSS and using `ExtractTextPlugin` to output a separate stylesheet, avoiding styles being blocked by JS and cutting the runtime overhead of parsing CSS.
* Leveraging route-based code splitting: so [Webpack can split bundles using the new proposed dynamic `import()` syntax](https://webpack.js.org/guides/code-splitting-async/). [TypeScript 2.4](https://blogs.msdn.microsoft.com/typescript/2017/06/27/announcing-typescript-2-4/) added support for `import()` and we can now defer loading huge routes until they are needed.
* Using [`svg-sprite-loader`](https://github.com/kisenka/svg-sprite-loader) instead of fetching the entire Font Awesome assets. I created a custom SVG sprite that included only the icons, this way I could mix and match icons from any popular icon font or icon set and I didn't need to import the entire set. [SVG is almost always a better choice compared to icon fonts](https://css-tricks.com/icon-fonts-vs-svg/) and has a [pretty good browser support](https://caniuse.com/#search=svg), going back to IE9.
* Instead of compiling to ES5 and including all ES6 polyfills, I used [`babel-preset-env`](https://github.com/babel/babel-preset-env) which is like Autoprefixer for JS: you specify a list of browsers you need to support, and it only compiles as needed. I also used [Babili](https://github.com/babel/babili/) instead of UglifyJS as the latter does not support ES6+ currently.
* Only including used functions from lodash and all other libraries, e.g. `import sortBy from 'lodash/sortBy'` instead of `import { sortBy } from 'lodash'`. Although Webpack tries to perform tree-shaking, the entire lodash library was being included, as Webpack cannot tree-shake CommonJS modules.

Applying these optimizations improved the entry bundle size, TTFB, and runtime performance significantly on mobile.

I also made most of these steps configurable by build flags, so I could easily compare using React and Preact and see how that could affect the bundle size, with something like that:

```
PREACT=1 yarn webpack/build # Bundle Preact instead of React
ES5=1 yarn webpack/build # Down-level compile to ES5 instead of using `babel-preset-env`
```

You can see how that is configured in [my Webpack config for WebCeph 2](https://github.com/forabi/WebCeph2/blob/aa34e45c6b4db8660dee3e1ec033b30d7621c167/webpack.config.js).

## Enforcing performance constraints
In addition to making these optimizations, we should prevent any regressions and make that a required step for a successful build. Webpack 2 can [enforce limits](https://webpack.js.org/configuration/performance) on the size of each output file, we can fail the build if any file is larger than a specified limit.

We should also integrate with local monitoring tools like [Lighthouse](https://github.com/addyosmani/webpack-lighthouse-plugin), and with remote ones like [WebPageTest](webpagetest.org) for realistic performance metrics.

## Relevant links
[Addy Osmani: The Browser Hackers Guide To Instantly Loading Everything | JSConf EU 2017
](https://www.youtube.com/watch?v=7vUs5yOuv-o&t=1445s)

**Comment by forabi** [comment]
I made some progress on this.
The productions bundle is now split across 3 chunks
1. `app.js` (38.6K) contains our own components and routes.
2. `react.js` (76.4K) contains `preact`, `preact-compat`, `react-router` and all other React libraries.
3. `vendor.js` (541.8K) contains other non-React-specific code, like Firebase and Algolia integrations.

There is also a separate `manifest.js` bundle that contains webpack runtime metadata, this should go in a separate bundle in order to leverage long term caching.

All chunks are hashed based on content and have persistent module names so they are reproducible with every build. If no changes are made to code in a specific chunk, this chunk won't be fetched again by browsers (provided we set up `Cache-Control` headers correctly`.

Next step is to analyze the `vendor.js` chunk, and see if we can improve. Looking at the new [`stats.json`](https://www.dropbox.com/s/pzkm5tac9lru1oj/stats-2.json?dl=0) file with [Webpack Visualizer](https://chrisbateman.github.io/webpack-visualizer/), we can already see areas for improvements. For example, we can only include the Firebase services we are using, this is already [documented](https://firebase.google.com/docs/web/setup).

![screenshot_20170629_220735](https://user-images.githubusercontent.com/1012761/27705974-7744dd58-5d18-11e7-94e3-86e2e561acaa.png)

After that we need to start looking at our routes and using React-Loadable with dynamic `import()` to generate route-based chunks. This way routes will only be loaded when (and if) needed.

I've also added [resource hints](https://css-tricks.com/prefetching-preloading-prebrowsing/) to preload relevant chunks asynchronously (`<link rel="preload">`).

**Comment by wholesomedev** [comment]
Awesome work @forabi! 

All of the changes that you're proposing seem good to me. I look forward to the final results. Feel free to send a PR whenever that's ready.

Regarding enforcement of performance constraints, another thing we can do is log the performance of the app, like how quickly it downloads, installs, and renders. (Metrics and logging is a whole different challenge, though. We can tackle that when we're ready.)

**Comment by forabi** [comment]
Chrome (and Firefox probably) support the Performance Observer API and
Navigation Timing API which give metrics about network and rendering times.
We could grab these metrics and track them periodically.

But I agree. We should probably focus on completing a usable v1 before
getting to that.

The PR is almost ready. There is some babel dependency missing in the
development build and I'm trying to figure out what it is.
